massa-event-cache/src/event_cache.rs

```diff
@@ -0,0 +1,454 @@
+        let db = DB::open_default(path).expect(OPEN_ERROR);
+        let entry_count = db.iterator(IteratorMode::Start).count();
+
+        Self {
+            db,
+            entry_count,
+            max_entry_count,
+            snip_amount,
+            event_ser: SCOutputEventSerializer::new(),
+            event_deser: SCOutputEventDeserializer::new(SCOutputEventDeserializerArgs {
+                thread_count,
+            }),
+        }
+    }
+
+    #[allow(dead_code)]
+    /// Insert a new event in the cache
+    pub fn insert(&mut self, event: SCOutputEvent) {
+        if self.entry_count >= self.max_entry_count {
+            self.snip(None);
+        }
+
+        let event_key = {
+            let mut event_key = event.context.slot.to_bytes_key().to_vec();
+            event_key.extend(event.context.index_in_slot.to_be_bytes());
+            event_key
+        };
+        let mut event_buffer = Vec::new();
+        self.event_ser.serialize(&event, &mut event_buffer).unwrap();
+
+        let mut batch = WriteBatch::default();
+        batch.put(event_key, event_buffer);
+        self.db.write(batch).expect(CRUD_ERROR);
+        self.entry_count = self.entry_count.saturating_add(1);
+
+        debug!("(Event insert) entry_count is: {}", self.entry_count);
+    }
+
+    /// Insert new events in the cache
+    ///
+    /// For performance reason, pass events_len to avoid cloning the iterator
+    pub fn insert_multi_it(
+        &mut self,
+        events: impl Iterator<Item = SCOutputEvent> + Clone,
+        events_len: Option<usize>,
+    ) {
+        let events_len = events_len.unwrap_or_else(|| events.clone().count());
+
+        if self.entry_count + events_len >= self.max_entry_count {
+            let snip_amount = max(self.snip_amount, events_len);
+            self.snip(Some(snip_amount));
+        }
+
+        let mut batch = WriteBatch::default();
+        for event in events {
+            let event_key = {
+                let mut event_key = event.context.slot.to_bytes_key().to_vec();
+                event_key.extend(event.context.index_in_slot.to_be_bytes());
+                event_key
+            };
+            let mut event_buffer = Vec::new();
+            self.event_ser.serialize(&event, &mut event_buffer).unwrap();
+
+            batch.put(event_key, event_buffer);
+        }
+        self.db.write(batch).expect(CRUD_ERROR);
+        self.entry_count = self.entry_count.saturating_add(events_len);
+
+        debug!("(Events insert) entry_count is: {}", self.entry_count);
+    }
+
+    fn db_iter(&self, mode: Option<IteratorMode>) -> DBIterator {
+        self.db.iterator(mode.unwrap_or(IteratorMode::Start))
+    }
+
+    pub(crate) fn get_filtered_sc_output_events<'b, 'a: 'b>(
+        &'a self,
+        filter: &'b EventFilter,
+    ) -> impl Iterator<Item = SCOutputEvent> + 'b {
+        self.db_iter(Some(IteratorMode::Start)).filter_map(|kvb| {
```

Comment:
With millions of events in cache, this function will always crash the node.

TODO:
* only iterate on selected slot ranges
* create indexes with respect to the available filters